File: frontend/src/lib/server/pmg-scheduler.ts
import { formatUserError } from "@/lib/format-api-error";
import { calcularPmg } from "@/lib/pmg";
import type { Lancamento, PmgResumo, TesourariaLancamento } from "@/lib/types";
import { enviarPmg, PmgDeliveryError } from "@/lib/server/pmg-delivery";
import { getSupabaseAdmin } from "@/lib/server/supabase-admin";

interface AgendamentoRow {
  id: string;
  empresa_id: string;
  canal: "email" | "whatsapp";
  destinatario: string;
  dia_semana: number;
  ativo: boolean;
  empresas: { nome: string } | { nome: string }[] | null;
}

function empresaNome(ag: AgendamentoRow): string {
  if (!ag.empresas) return "Empresa";
  if (Array.isArray(ag.empresas)) return ag.empresas[0]?.nome ?? "Empresa";
  return ag.empresas.nome ?? "Empresa";
}

function previousMonthPeriod(reference = new Date()) {
  const ref = new Date(reference.getFullYear(), reference.getMonth(), 1);
  ref.setMonth(ref.getMonth() - 1);
  const year = ref.getFullYear();
  const month = ref.getMonth() + 1;
  const periodo = `${year}-${String(month).padStart(2, "0")}`;
  const inicio = `${periodo}-01`;
  const lastDay = new Date(year, month, 0).getDate();
  const fim = `${periodo}-${String(lastDay).padStart(2, "0")}`;
  return { periodo, inicio, fim };
}

function periodBounds(periodo: string) {
  const [year, month] = periodo.split("-").map(Number);
  const inicio = `${periodo}-01`;
  const lastDay = new Date(year, month, 0).getDate();
  const fim = `${periodo}-${String(lastDay).padStart(2, "0")}`;
  return { periodo, inicio, fim };
}

async function fetchActiveAgendamentos(diaSemana?: number) {
  const supabase = getSupabaseAdmin();
  let query = supabase
    .from("pmg_agendamentos")
    .select("id, empresa_id, canal, destinatario, dia_semana, ativo, empresas(nome)")
    .eq("ativo", true);

  if (diaSemana !== undefined) {
    query = query.eq("dia_semana", diaSemana);
  }

  const { data, error } = await query;
  if (error) throw new Error(formatUserError(error.message, "Erro ao buscar agendamentos."));
  return (data ?? []) as AgendamentoRow[];
}

async function alreadySentThisWeek(empresaId: string, canal: string): Promise<boolean> {
  const supabase = getSupabaseAdmin();
  const weekAgo = new Date(Date.now() - 6 * 24 * 60 * 60 * 1000).toISOString();
  const { data } = await supabase
    .from("envios_pmg")
    .select("id")
    .eq("empresa_id", empresaId)
    .eq("canal", canal)
    .eq("status", "enviado")
    .gte("created_at", weekAgo);

  return (data?.length ?? 0) > 0;
}

async function fetchPmgData(empresaId: string, inicio: string, fim: string) {
  const supabase = getSupabaseAdmin();
  const [lancamentos, tesouraria] = await Promise.all([
    supabase
      .from("lancamentos")
      .select("valor, natureza, categoria, categoria_corrigida, data")
      .eq("empresa_id", empresaId)
      .gte("data", inicio)
      .lte("data", fim),
    supabase
      .from("tesouraria_lancamentos")
      .select("valor, natureza, categoria, data")
      .eq("empresa_id", empresaId)
      .gte("data", inicio)
      .lte("data", fim),
  ]);

  if (lancamentos.error) {
    throw new Error(formatUserError(lancamentos.error.message, "Erro ao buscar lançamentos."));
  }
  if (tesouraria.error) {
    throw new Error(formatUserError(tesouraria.error.message, "Erro ao buscar tesouraria."));
  }

  return {
    lancamentos: (lancamentos.data ?? []) as Lancamento[],
    tesouraria: (tesouraria.data ?? []) as TesourariaLancamento[],
  };
}

export async function runWeeklyPmgDispatch(options?: {
  force?: boolean;
  periodo?: string;
}) {
  const force = options?.force ?? false;
  const today = new Date();
  const weekday = today.getDay() === 0 ? 6 : today.getDay() - 1; // seg=0 … dom=6

  const agendamentos = force
    ? await fetchActiveAgendamentos()
    : await fetchActiveAgendamentos(weekday);

  const { periodo, inicio, fim } = options?.periodo
    ? periodBounds(options.periodo)
    : previousMonthPeriod(today);
  const supabase = getSupabaseAdmin();

  const enviados: Array<{ empresa: string; destinatario: string; canal: string; periodo: string }> =
    [];
  const erros: Array<{ empresa: string; destinatario: string; erro: string }> = [];
  const ignorados: Array<{ empresa_id: string; motivo: string }> = [];

  for (const ag of agendamentos) {
    const empresaNomeVal = empresaNome(ag);

    if (!force && (await alreadySentThisWeek(ag.empresa_id, ag.canal))) {
      ignorados.push({
        empresa_id: ag.empresa_id,
        motivo: "Já enviado nos últimos 7 dias",
      });
      continue;
    }

    const { lancamentos, tesouraria } = await fetchPmgData(ag.empresa_id, inicio, fim);
    if (!lancamentos.length && !tesouraria.length) {
      ignorados.push({
        empresa_id: ag.empresa_id,
        motivo: `Sem lançamentos em ${periodo}`,
      });
      continue;
    }

    const resumo: PmgResumo = calcularPmg(lancamentos, tesouraria);

    try {
      await enviarPmg({
        empresa_id: ag.empresa_id,
        empresa_nome: empresaNomeVal,
        periodo,
        periodo_inicio: inicio,
        periodo_fim: fim,
        canal: ag.canal,
        destinatario: ag.destinatario,
        resumo,
        enviado_por: null,
      });

      await supabase
        .from("pmg_agendamentos")
        .update({ ultimo_envio_em: new Date().toISOString() })
        .eq("id", ag.id);

      enviados.push({
        empresa: empresaNomeVal,
        destinatario: ag.destinatario,
        canal: ag.canal,
        periodo,
      });
    } catch (err) {
      const message = err instanceof PmgDeliveryError ? err.message : String(err);
      erros.push({
        empresa: empresaNomeVal,
        destinatario: ag.destinatario,
        erro: message,
      });
    }
  }

  return {
    periodo,
    dia_semana: weekday,
    total_agendamentos: agendamentos.length,
    enviados,
    erros,
    ignorados,
  };
}

File: backend/src/acc_nomad/services/pmg_scheduler.py
"""Disparo semanal automático do PMG."""

from __future__ import annotations

import calendar
from datetime import UTC, date, datetime, timedelta
from typing import Any

from acc_nomad.services.pmg_calculator import calcular_pmg_resumo, fetch_pmg_data
from acc_nomad.services.pmg_delivery import DeliveryError, enviar_pmg
from acc_nomad.services.supabase_client import get_supabase_client


def previous_month_period(reference: date | None = None) -> tuple[str, date, date]:
    ref = reference or date.today()
    if ref.month == 1:
        year, month = ref.year - 1, 12
    else:
        year, month = ref.year, ref.month - 1
    inicio = date(year, month, 1)
    fim = date(year, month, calendar.monthrange(year, month)[1])
    periodo = f"{year:04d}-{month:02d}"
    return periodo, inicio, fim


def fetch_active_agendamentos(dia_semana: int | None = None) -> list[dict[str, Any]]:
    client = get_supabase_client()
    query = client.table("pmg_agendamentos").select("*, empresas(nome)").eq("ativo", True)
    if dia_semana is not None:
        query = query.eq("dia_semana", dia_semana)
    result = query.execute()
    return result.data or []


def _already_sent_this_week(client, empresa_id: str, periodo: str, canal: str) -> bool:
    week_ago = (datetime.now(UTC) - timedelta(days=6)).isoformat()
    result = (
        client.table("envios_pmg")
        .select("id")
        .eq("empresa_id", empresa_id)
        .eq("canal", canal)
        .eq("status", "enviado")
        .gte("created_at", week_ago)
        .execute()
    )
    rows = result.data or []
    return len(rows) > 0


async def run_weekly_pmg_dispatch(
    *,
    force: bool = False,
    dia_semana: int | None = None,
) -> dict[str, Any]:
    today = date.today()
    weekday = dia_semana if dia_semana is not None else today.weekday()

    if force:
        agendamentos = fetch_active_agendamentos()
    else:
        agendamentos = fetch_active_agendamentos(dia_semana=weekday)

    periodo, inicio, fim = previous_month_period(today)
    client = get_supabase_client()

    enviados: list[dict[str, str]] = []
    erros: list[dict[str, str]] = []
    ignorados: list[dict[str, str]] = []

    for ag in agendamentos:
        empresa_id = ag["empresa_id"]
        canal = ag["canal"]
        destinatario = ag["destinatario"]
        empresa_nome = (ag.get("empresas") or {}).get("nome") or "Empresa"

        if not force and _already_sent_this_week(client, empresa_id, periodo, canal):
            ignorados.append(
                {
                    "empresa_id": empresa_id,
                    "motivo": "Já enviado nos últimos 7 dias",
                }
            )
            continue

        lancamentos, tesouraria = fetch_pmg_data(empresa_id, inicio, fim)
        resumo = calcular_pmg_resumo(lancamentos, tesouraria)

        if not lancamentos and not tesouraria:
            ignorados.append(
                {
                    "empresa_id": empresa_id,
                    "motivo": f"Sem lançamentos em {periodo}",
                }
            )
            continue

        try:
            await enviar_pmg(
                empresa_id=empresa_id,
                empresa_nome=empresa_nome,
                periodo=periodo,
                periodo_inicio=inicio,
                periodo_fim=fim,
                canal=canal,
                destinatario=destinatario,
                resumo=resumo,
                enviado_por=None,
            )
            client.table("pmg_agendamentos").update(
                {"ultimo_envio_em": datetime.now(UTC).isoformat()}
            ).eq("id", ag["id"]).execute()
            enviados.append(
                {
                    "empresa": empresa_nome,
                    "destinatario": destinatario,
                    "canal": canal,
                    "periodo": periodo,
                }
            )
        except DeliveryError as exc:
            erros.append(
                {
                    "empresa": empresa_nome,
                    "destinatario": destinatario,
                    "erro": str(exc),
                }
            )

    return {
        "periodo": periodo,
        "dia_semana": weekday,
        "total_agendamentos": len(agendamentos),
        "enviados": enviados,
        "erros": erros,
        "ignorados": ignorados,
    }
